In [2]:
import tensorflow as tf
from tensorflow.keras.layers import Dense
import numpy as np
import wandb
from wandb.keras import WandbCallback

In [ ]:
wandb.init(project="Boston Housing")

In [ ]:
config = wandb.config
config.learning_rate = 0.001

### Data

In [3]:
boston_dataset = tf.keras.datasets.boston_housing
(X_train, Y_train), (X_test, Y_test) = boston_dataset.load_data()

In [4]:
Y_train = Y_train.reshape(-1, 1)
print(X_train.shape, Y_train.shape)

(404, 13) (404, 1)


In [5]:
Y_test = Y_test.reshape(-1, 1)
print(X_test.shape, Y_test.shape)

(102, 13) (102, 1)


## Model

In [9]:
model = tf.keras.models.Sequential([
    Dense(24, input_dim=13, activation='sigmoid'), # Hidden Layer 1 # input_dim moshakhas konande input layer ast
    Dense(8, activation='sigmoid'), # HL2
    Dense(4, activation='sigmoid'), # HL3
    Dense(1, activation='linear') # Output Layer
])

In [10]:
model.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=config.learning_rate),
    loss=tf.keras.losses.MeanSquaredError()
)

In [11]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 8)                 112       
                                                                 
 dense_5 (Dense)             (None, 8)                 72        
                                                                 
 dense_6 (Dense)             (None, 4)                 36        
                                                                 
 dense_7 (Dense)             (None, 1)                 5         
                                                                 
Total params: 225
Trainable params: 225
Non-trainable params: 0
_________________________________________________________________


- param laye aval = (13 * 8) w + 8 b = 112
- param laye badi = (8 * 8) w + 8 b = 72
- param laye badi = (8 * 4) w + 4 = 36
- param laye akhar = 4 ta w darim va yeki bias = 5

## Train

In [12]:
model.fit(X_train, Y_train, epochs=1000, callbacks=[WandbCallback()])

Epoch 1/32
13/13 [==============================] - 2s 6ms/step - loss: 550.9857
Epoch 2/32
13/13 [==============================] - 0s 6ms/step - loss: 493.3609
Epoch 3/32
13/13 [==============================] - 0s 3ms/step - loss: 437.3031
Epoch 4/32
13/13 [==============================] - 0s 3ms/step - loss: 384.7303
Epoch 5/32
13/13 [==============================] - 0s 5ms/step - loss: 337.5227
Epoch 6/32
13/13 [==============================] - 0s 3ms/step - loss: 296.5101
Epoch 7/32
13/13 [==============================] - 0s 4ms/step - loss: 260.7116
Epoch 8/32
13/13 [==============================] - 0s 3ms/step - loss: 229.3204
Epoch 9/32
13/13 [==============================] - 0s 2ms/step - loss: 202.0858
Epoch 10/32
13/13 [==============================] - 0s 5ms/step - loss: 178.6416
Epoch 11/32
13/13 [==============================] - 0s 2ms/step - loss: 158.7946
Epoch 12/32
13/13 [==============================] - 0s 2ms/step - loss: 142.5462
Epoch 13/32
13/13 [======

akharin loss age 84 hastesh bayad bekhater MSE jazresh ro begirim va dar nahayat yani dar har qeymat khone hodod 9 hezar dollar darim eshtebah migim

## Evaluate

In [ ]:
model.evaluate(X_test, Y_test)

## Inference

In [ ]:
my_house = np.array([[0.00632, 18.0, 2.31, 0.0, 0.538, 6.575, 65.2, 4.0900, 1.0, 296.0, 15.3, 396.90, 4.98]])

model.predict(my_house)